xds/internal/client/lds_test.go

```diff
@@ -933,8 +990,428 @@ func (s) TestUnmarshalListener_ServerSide(t *testing.T) {
+				FilterChainMatch: &v3listenerpb.FilterChainMatch{
+					SourcePrefixRanges: []*v3corepb.CidrRange{
+						{
+							AddressPrefix: "a.b.c.d",
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			desc: "bad source prefix length",
+			inputFilterChain: &v3listenerpb.FilterChain{
+				FilterChainMatch: &v3listenerpb.FilterChainMatch{
+					SourcePrefixRanges: []*v3corepb.CidrRange{
+						{
+							AddressPrefix: "10.1.1.0",
+							PrefixLen: &wrapperspb.UInt32Value{
+								Value: 50,
+							},
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			desc: "source prefix ranges",
+			inputFilterChain: &v3listenerpb.FilterChain{
+				FilterChainMatch: &v3listenerpb.FilterChainMatch{
+					SourcePrefixRanges: []*v3corepb.CidrRange{
+						{
+							AddressPrefix: "10.1.1.0",
+							PrefixLen: &wrapperspb.UInt32Value{
+								Value: 8,
+							},
+						},
+						{
+							AddressPrefix: "192.168.1.0",
+							PrefixLen: &wrapperspb.UInt32Value{
+								Value: 24,
+							},
+						},
+					},
+				},
+			},
+			wantFilterChain: &FilterChain{
+				Match: &FilterChainMatch{
+					SourcePrefixRanges: []net.IP{
+						net.IPv4(10, 1, 1, 0),
+						net.IPv4(192, 168, 1, 0),
+					},
+				},
+			},
+		},
+		{
+			desc:             "empty transport socket",
+			inputFilterChain: &v3listenerpb.FilterChain{},
+			wantFilterChain:  &FilterChain{Match: &FilterChainMatch{}},
+		},
+		{
+			desc: "bad transport socket name",
+			inputFilterChain: &v3listenerpb.FilterChain{
+				TransportSocket: &v3corepb.TransportSocket{
+					Name: "unsupported-transport-socket-name",
+				},
+			},
+			wantErr: true,
+		},
+		{
+			desc: "unexpected url in transport socket",
+			inputFilterChain: &v3listenerpb.FilterChain{
+				TransportSocket: &v3corepb.TransportSocket{
+					Name: "envoy.transport_sockets.tls",
+					ConfigType: &v3corepb.TransportSocket_TypedConfig{
+						TypedConfig: &anypb.Any{
+							TypeUrl: version.V3UpstreamTLSContextURL,
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			desc: "badly marshaled downstream tls context",
+			inputFilterChain: &v3listenerpb.FilterChain{
+				TransportSocket: &v3corepb.TransportSocket{
+					Name: "envoy.transport_sockets.tls",
+					ConfigType: &v3corepb.TransportSocket_TypedConfig{
+						TypedConfig: &anypb.Any{
+							TypeUrl: version.V3DownstreamTLSContextURL,
+							Value:   []byte{1, 2, 3, 4},
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			desc: "missing common tls context",
+			inputFilterChain: &v3listenerpb.FilterChain{
+				TransportSocket: &v3corepb.TransportSocket{
+					Name: "envoy.transport_sockets.tls",
+					ConfigType: &v3corepb.TransportSocket_TypedConfig{
+						TypedConfig: &anypb.Any{
+							TypeUrl: version.V3DownstreamTLSContextURL,
+							Value: func() []byte {
+								tls := &v3tlspb.DownstreamTlsContext{}
+								mtls, _ := proto.Marshal(tls)
+								return mtls
+							}(),
+						},
```

Comment:
Let's use ptypes.MarshalAny instead, otherwise we're trusting our constant we set in the type_url field: https://pkg.go.dev/github.com/adnaan/protobuf/ptypes#MarshalAny

(and below)